sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -184,7 +183,8 @@ class Dataset[T] private[sql](
       // to happen right away to let these side effects take place eagerly.
       case p if hasSideEffects(p) =>
         LogicalRDD(queryExecution.analyzed.output, queryExecution.toRdd)(sparkSession)
-      case Union(children) if children.forall(hasSideEffects) =>
+      case Union(children) if children.forall(x =>
+          x.isInstanceOf[InsertIntoTable] || x.isInstanceOf[InsertIntoHadoopFsRelationCommand]) =>
```

Comment:
1. If you want to allow union query containing side effect, 'forall' should be replaced 'exists'. 
2. If you want to allow multi-insert syntax only(not union query), we should do this.

I think 2 was your intention before.